DSCResources/MSFT_xDiskAccessPath/MSFT_xDiskAccessPath.psm1

```diff
@@ -227,77 +222,139 @@ function Set-TargetResource
+            {
+                # Find the first basic partition of any size
+                $partition = $partition |
+                    Where-Object -Filter { $_.Type -eq 'Basic' } |
+                    Select-Object -First 1
+            } # if
 
-        # Create the partition.
-        $partition = New-Partition @partitionParams
+            if ($partition)
+            {
+                # A partition matching the required size was found
+                Write-Verbose -Message ($LocalizedData.MatchingPartitionFoundMessage -f `
+                    $DiskNumber,$partition.PartitionNumber)
 
-        # After creating the partition it can take a few seconds for it to become writeable
-        # Wait for up to 30 seconds for the parition to become writeable
-        $timeout = 30000
-        $start = [DateTime]::Now
-        While ($partition.IsReadOnly `
-            -and ([DateTime]::Now - $start).TotalMilliseconds -lt $timeout)
+                $createPartition = $false
+            }
+            else
+            {
+                # A partition matching the required size was not found
+                Write-Verbose -Message ($LocalizedData.MatchingPartitionNotFoundMessage -f `
+                    $DiskNumber)
+            } # if
+        } # if
+
+        # Do we need to create a new partition?
+        if ($createPartition)
         {
-            Write-Verbose -Message ($LocalizedData.NewPartitionIsReadOnlyMessage -f `
-                $partition.DiskNumber,$partition.PartitionNumber)
+            # Attempt to create a new partition
+            $partitionParams = @{
+                DiskNumber = $DiskNumber
+            }
 
-            Start-Sleep -Seconds 1
+            if ($Size)
+            {
+                # Use only a specific size
+                Write-Verbose -Message ( @(
+                        "$($MyInvocation.MyCommand): "
+                        $($LocalizedData.CreatingPartitionMessage `
+                            -f $DiskNumber,"$($Size/1KB) KB")
+                    ) -join '' )
+                $partitionParams["Size"] = $Size
+            }
+            else
+            {
+                # Use the entire disk
+                Write-Verbose -Message ( @(
+                        "$($MyInvocation.MyCommand): "
+                        $($LocalizedData.CreatingPartitionMessage `
+                            -f $DiskNumber,'all free space')
+                    ) -join '' )
+                $partitionParams["UseMaximumSize"] = $true
+            } # if
+
+            # Create the partition
+            $partition = New-Partition @partitionParams
+
+            # After creating the partition it can take a few seconds for it to become writeable
+            # Wait for up to 30 seconds for the parition to become writeable
+            $timeout = 30000
+            $start = [DateTime]::Now
+            While ($partition.IsReadOnly `
+                -and ([DateTime]::Now - $start).TotalMilliseconds -lt $timeout)
```

Comment:
Minor: You are creating a time span each iteration of the loop.  Calculate the timeout as an absolute time and check if `now` is still less than the calculated time.

Minor: you should use Get-Date instead of `[DateTime]::Now`